src/net_processing.cpp

```diff
@@ -657,25 +663,56 @@ static void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vec
     }
 }
 
-void RequestTx(CNodeState *state, uint256 &txid) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+void EraseTxRequest(uint256& txid) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    m_already_asked_for.erase(txid);
+}
+
+int64_t GetTxRequestTime(const uint256& txid)
+{
+    auto it = m_already_asked_for.find(txid);
+    if (it != m_already_asked_for.end()) {
+        return it->second;
+    }
+    return 0;
+}
+
+void UpdateTxRequestTime(const uint256 &txid, int64_t request_time)
 {
-    CNodeState::TxDownloadState &peer = state->m_tx_download;
-    if (peer.m_tx_announced.size() >= MAX_PEER_TX_ANNOUNCEMENTS || peer.m_tx_announced.count(txid)) {
+    auto it = m_already_asked_for.find(txid);
+    if (it == m_already_asked_for.end()) {
+        m_already_asked_for.insert(std::make_pair(txid, request_time));
+    } else {
+        m_already_asked_for.update(it, request_time);
+    }
+}
+
+
+void RequestTx(CNodeState *state, const uint256 &txid) {
+    CNodeState::TxDownloadState &peer_download_state = state->m_tx_download;
+    if (peer_download_state.m_tx_announced.size() >= MAX_PEER_TX_ANNOUNCEMENTS || peer_download_state.m_tx_announced.count(txid)) {
         // Too many queued announcements from this peer, or we already have
         // this announcement
         return;
     }
-    peer.m_tx_announced.insert(txid);
 
-    int64_t process_time = GetTimeMicros();
+    int64_t process_time;
+    int64_t last_request_time = GetTxRequestTime(txid);
+    // First time requesting this tx
+    if (last_request_time == 0) {
+        peer_download_state.m_tx_announced.insert(txid);
+        process_time = GetTimeMicros();
+    } else {
+        // Add extra delay so that first peer from the list is not prioritizied due to ThreadMessageHandler delay
+        process_time = PoissonNextSend(last_request_time + MAX_GETDATA_TX_DELAY, AVG_GETDATA_REQUEST_INTERVAL);
```

Comment:
There's no downside here to using a uniform distribution here, is there?  And otherwise some inbound peers could be prioritized over outbound peers, which is what we were trying to avoid... I think we could just do:

`process_time = last_request_time + MAX_GETDATA_TX_DELAY + GetRand(AVG_GETDATA_REQUEST_INTERVAL * 1000000)`

Reply:
> some inbound peers could be prioritized over outbound peers

Nope, with these constants, it's [60..62] for outbound and [62..64] for inbound, so it's not possible.

However, I think you're right and using GetRand() would work.

Reply:
> Nope, with these constants, it's [60..62] for outbound and [62..64] for inbound, so it's not possible.

There is no upper bound on what PoissonNextSend might return...?

Reply:
You're right, thanks.